### PR TITLE
Wallet Performance Improvements

### PR DESCRIPTION
**Pull request by ianadavies**
This PR makes a number of improvements for Wallet Performance.

Have improved constructor of `uint256`. This looks a little messy but we create 100's of 000's of these things and we were allocating like crazy (3 arrays just to initialise) the new constructor does not allocate saving the GC a lot of work.

Also improved `GetHistory()` on the `WalletController` by ~3x. By using lookups on `txId`

### CODE REVIEW

File: src/NBitcoin/UInt256.cs [hunk 1]
```diff
@@ -275,8 +289,11 @@ public override bool Equals(object obj)
             var item = obj as uint256;
             if (item == null)
                 return false;
+            
+            if (this.pn0 != item.pn0)
```
Comment (ianadavies) [inline src/NBitcoin/UInt256.cs]: I decided to put this in as its 5% performance improvement in release mode and 30% in debug

File: src/NBitcoin/UInt256.cs [hunk 2]
```diff
@@ -251,18 +251,32 @@ public uint256(string str)
             if (str.StartsWith("0x", StringComparison.OrdinalIgnoreCase))
                 str = str.Substring(2);
 
-            byte[] bytes = Encoder.DecodeData(str).Reverse().ToArray();
```
Comment (ianadavies) [inline src/NBitcoin/UInt256.cs]: We were creating 3 arrays here! 2 in this class and 1 in encoder

File: src/NBitcoin/UInt256.cs (hunk 1 above)
Comment (ArcSin2000X) [reply src/NBitcoin/UInt256.cs]: how does that translate to actual difference it time?

Comment (ianadavies) [reply src/NBitcoin/UInt256.cs]: Negligible in real world 5ms on 1m comparisons, but debug it was a big difference ~40ms so useful while developing I guess

File: src/NBitcoin/UInt256.cs [hunk 3]
```diff
@@ -251,18 +251,32 @@ public uint256(string str)
...
+                    enumerator.MoveNext();
+                    this.pn2 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn1 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn0 = enumerator.Current;
+                }
+                catch(InvalidOperationException)
+                {
+                    throw new FormatException("Invalid hex length");
```
Comment (rowandh) [inline src/NBitcoin/UInt256.cs]: Nit: this changes functionality as enumerables with more than 8 elements are now possible but there is no longer a check for this.

Comment (ianadavies) [reply src/NBitcoin/UInt256.cs]: Hmm good spot.

Could do this

` try
                {
                    enumerator.MoveNext();
                    this.pn7 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn6 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn5 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn4 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn3 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn2 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn1 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn0 = enumerator.Current;
                    if (enumerator.MoveNext())
                    {
                        throw new InvalidOperationException();
                    }
                }
                catch(InvalidOperationException)
                {
                    throw new FormatException("Invalid hex length");
                }`

**Comment by rowandh** [comment]
IMO would be better to PR the uint changes separately as they're much broader than just wallet improvements and it will be easier to review etc. in isolation.

**Comment by MithrilMan** [comment]
all improvements in this section should be compared on upstream NBitcoin, btw we'll get way more improvement using Span<T> classes from dotnetcore 2.2+ (we had already an open issue about this upgrade)

**Comment by quantumagi** [comment]
```
  X GetHistoryWithChangeAddressesShouldIncludeSpentChangeAddesses [25ms]
  X GetHistoryWithValidModelWithTransactionSpendingDetailsReturnsWalletHistoryModel 
```

**Comment by ianadavies** [comment]
> ```
>   X GetHistoryWithChangeAddressesShouldIncludeSpentChangeAddesses [25ms]
>   X GetHistoryWithValidModelWithTransactionSpendingDetailsReturnsWalletHistoryModel 
> ```

**Comment by ianadavies** [comment]
> ```
>   X GetHistoryWithChangeAddressesShouldIncludeSpentChangeAddesses [25ms]
>   X GetHistoryWithValidModelWithTransactionSpendingDetailsReturnsWalletHistoryModel 
> ```
These pass locally for me. I think build server is playing up

**Comment by dangershony** [comment]
The name of this PR is totally misleading, changing uint256 is the heart of the code base to improve history sounds like a bad treadoff, as @MithrilMan said there are improvements to uin256 in NBitcoin to use spans and array pools.

**Comment by codingupastorm** [comment]
Commenting only on the uint changes:

In most situations I'd be for this, but with a release inbound unfortunately I agree with the others:
- It affects more things than just the wallet
- We will be able to adopt the NBitcoin solution soon anyway
- Micro-optimisations like this should be the last part of the optimisation process. My guess is this could make a 750ms GetHistory call to be 700ms. What we really need to do is get it to 70...

**Comment by ArcSin2000X** [comment]
Change to be part of 307 scope

**Comment by quantumagi** [comment]
Risk/reward maybe not worth it at this time. Rather stick to the proven albeit sightly slower code.

**Pull request closed without merging**